### ISSUE TITLE
关于Pig Latin的seq2seq model in

### ISSUE THREAD

**Issue by dodocho** [Linusp/soph#1]
你好，谢谢这个帖子。有一个问题想请教一下，文章中说“A、B、C 和表示序列结束的特殊符号 <EOS> 依次输入，并在读取到 <EOS> 时终止接收输入”，然后我check了您的pig_latin.py文件，发现每一个chracter您用28维的one—hot向量表达，比如“a”这个字母用［0，0，1，0，0，0，...， 0］表达，这个前两个位置分别代表，END_SYMBOL 和START_SYMBOL。对于一个单词首位用START_SYMBO即，［0，1，0，0，0，..., 0］，不足的用［1，0，0，0, ..., 0 ］ 补全。

我的问题是：（1）假如不用这个END_SYMBOL 和START_SYMBOL，会对结果有什么影响吗？（当然还是要用另外的符号讲句子填充到定长）。想这篇文章的第一幅图里面，EOS符号是表示读取结束的意思，但是如果用keras.model.fit 来训练, 似乎酒不必添加这个了，因为输入序列和输出序列是分开的，model.fit(x_train, y_trian)。谢谢

**Comment by Linusp** [comment]
@jimmyangc START_SYMBOL 其实并不是必要的，但 EOS 还是要填充的。添加这个符号的原因是因为模型要求训练时所有数据都是定长的，否则训练会出错，但在预测时，模型能够正确处理变长输入(前提是定义模型时输入层的只设置 input_dim 而不是设置 input_shape)

**Comment by Linusp** [comment]
@jimmyangc

**Comment by dodocho** [comment]
@Linusp 谢谢回复。那其实在keras中用encoder 和decoder 处理 seq to seq的问题时，只需要将输入序列(x_train)和输出序列(y_train)都填充为定长的，然后用model.fit(x_train, y_train)，模型就会自动编码和解码了？

**Comment by Linusp** [comment]
@jimmyangc 看下面这段代码

``` python
def build_model(input_size, seq_len, hidden_size):
    """建立一个 sequence to sequence 模型"""
    model = Sequential()
    model.add(GRU(input_dim=input_size, output_dim=hidden_size, return_sequences=False))
    model.add(Dense(hidden_size, activation="relu"))
    model.add(RepeatVector(seq_len))
    model.add(GRU(hidden_size, return_sequences=True))
    model.add(TimeDistributed(Dense(output_dim=input_size, activation="linear")))
    model.compile(loss="mse", optimizer='adam')

    return model
```

里面有两个 GRU ，第一个的 `return_sequences` 参数设置为 `False` 只取最后一个 time step 的输出，就起到了 encode 的作用，第二个 GRU 的 `return_sequences` 参数设置为 `True` ，就起到了 decode 的作用。

**Comment by Linusp** [comment]
@jimmyangc 当然了，我的代码里实现的其实是一个简化版的 seq2seq 模型，decoder 的每个 time step 的输入都是一样的，不像论文里提到的那样将 decoder 的上一次 time step 的输出作为下一个 time step 的输入，因此效果也就比较一般，勉强能处理像 Pig Latin 这样的简单问题。

完整的 seq2seq 模型可以参考 [farizrahmn4u/seq2seq](https://github.com/farizrahman4u/seq2seq)，里面还实现了带 attention 机制的 seq2seq 模型

**Comment by dodocho** [comment]
@Linusp 多谢明白了。如果将 Decoder 的每一个时刻的输出作为下一个时刻的输入，不知道这个代码片段decoder部分的循环是不是干这个的：

``` python
model = Sequential()
model.add(Embedding(input_dim=(len(chars)), output_dim=256, input_length=maxlen))

#encoding portion
model.add(LSTM(hidden_variables, return_sequences=True, input_shape=(maxlen, len(chars))))
model.add(Dropout(dropout))

model.add(RepeatVector(number_of_timesteps)

#decoding portion
for z in range(0,number_of_layers-2):
    model.add(LSTM(hidden_variables, return_sequences=True))
    model.add(Dropout(dropout))
model.add(LSTM(hidden_variables, return_sequences=False))
model.add(Dropout(dropout))
model.add(Dense(len(chars))) #this is the length of your output array. 
model.add(Activation('softmax')) 
model.compile(loss='mse', optimizer='adam')
```

**Comment by jimmyangc** [comment]
抱歉，上一段回复没有缩进
for z in range(0,number_of_layers-2):
      model.add(LSTM(hidden_variables, return_sequences=True))
      model.add(Dropout(dropout))

**Comment by jimmyangc** [comment]
抱歉上一段回复没有缩进：

for z in range(0,number_of_layers-2):
缩进缩进model.add(LSTM(hidden_variables, return_sequences=True))
缩进缩进model.add(Dropout(dropout))

**Comment by Linusp** [comment]
@jimmyangc 抱歉，不知道你这个代码片段哪来的，然后想靠循环来达成这个目的是不行的，“每个 time step 的输出作为下一个 time step 的输入” 是针对同一个 RNN layer(LSTM/GRU) 而言的，增加多个 LSTM layer 只能增强模型的表达能力，并不能实现 Decoder 的功能。

参考: [LSTMDecoder](https://github.com/farizrahman4u/seq2seq/blob/master/seq2seq/layers/decoders.py#L18)

**Comment by Linusp** [comment]
@jimmyangc 以及，缩进什么的用 Markdown 的 source block

**Comment by Linusp** [comment]
@jimmyangc 也就是说，如果要实现完整的 Decoder ，需要修改 LSTM/GRU 的实现，这个还是有点麻烦的，所以像我这个项目里的 Pig Latin 这样的问题，就偷个懒用简化版的了 :expressionless:

**Comment by dodocho** [comment]
@Linusp 谢谢回复。您那个pig latin作为一个introduction是简单易懂的范例，很不错的。但是看来用keras实现真正的seq2seq要啃啃 LSTM/GRU的代码了。[https://github.com/farizrahman4u/seq2seq](url) 好像很受欢迎，应该学习下！：)

**Comment by Linusp** [comment]
@jimmyangc :smile: :smile: :smile:

**Comment by Linusp** [comment]
@jimmyangc 另外，Keras 项目中有几个 Issue 讨论 seq2seq 模型，非常有阅读的价值，有时间的话建议去看一下
- [Keras·issue#395](https://github.com/fchollet/keras/issues/395)
- [Keras·issue#957](https://github.com/fchollet/keras/issues/957)
- [Keras·issue#2403](https://github.com/fchollet/keras/issues/2403)

需要注意因为时间问题，里面一些代码片段使用的是旧版 Keras 中的接口

**Comment by dodocho** [comment]
@Linusp  好的！ thank you so much！
